### ISSUE TITLE
[new-backend] Merging into master 

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#4485]
- [ ] basics of kdb framework exist, at least kdb get/set/rm works (@hannes99)
- [ ] basics of kdb mount work (@flo91)
- [x] hooks: at least gopts, spec & dbus basically works (@atmaxinger)
- [ ] Docs should be complete enough for newcomers to understand what to expect, what is broken (@kodebach)
- [x] Rebase onto `master` or merge `master` changes into `new-backend` (@kodebach)
- [ ] #4487 basic tests and reviews of everything above, either fix or disable all tests that are currently broken+document which is disabled in doc/todo/TESTING (@Eiskasten)

@hannes99 @flo91 @atmaxinger @kodebach @Eiskasten does this sound realistic for you to do in the next two weeks?

**Comment by markus2330** [comment]
Or to be more specific: What do you all think of doing a release of 0.9.11 on 4.10.2022 (@mpranj) and then merge the new-backend directly afterwards (@kodebach, but obviously all the tasks above must be fulfilled before by @hannes99 @flo91 @atmaxinger @Eiskasten).

**Comment by kodebach** [comment]
I added the most important task (rebase/merge from `master`)....

I think we also need a bit more details to decide, if this is feasible:

- The basics of `kdb` as well as `kdb mount` already work on the `new-backend` branch. They are still in C++ and the way `kdb mount` is implemented is not ideal. It makes a few assumptions that aren't necessary anymore, but when I fixed it, my intention was to stay as close to the behaviour of the old `kdb mount` as possible. What else is needed?
- The hook code for `gopts` is already merged and unless #4471 broke something, `spec` should also work as much as it does on `master`. Is removing `list` and calling `spec` directly enough, or should more be done to `spec` before the merge?
- I'm not really sure what you want me to do with the docs. Want I really don't want to do is write lots of documentation for this temporary state that will be outdated in a few months. That would just be waste of time. IMO the better solution would be to write documentation for how things should be when `new-backend` is released. Anything that isn't done yet should just be marked in the docs with a small TODO-like note. If more than that is needed, e.g. because you need to use a temporary workaround, I would just document that in GitHub issues. If there is still confusion, people should just ask.

**Comment by markus2330** [comment]
>  The basics of kdb as well as kdb mount already work on the new-backend branch. They are still in C++ and the way kdb mount is implemented is not ideal. It makes a few assumptions that aren't necessary anymore, but when I fixed it, my intention was to stay as close to the behaviour of the old kdb mount as possible. What else is needed?

- libmount in C
- kdb should use gopts

> Is removing list and calling spec directly enough, or should more be done to spec before the merge?

As discussed with @atmaxinger there should be tests for all the hooks.

> If there is still confusion, people should just ask.

This is okay to some extend. But the confusion must be at some later point so that people can at least successfully get to a point where they can ask a useful question (and also have the motivation to do so).

**Comment by kodebach** [comment]
> libmount in C

Why is this needed for the merge?

The current mounting code in `new-backend` works well enough for most test cases (including `testshell_markdown_*` and `testscr_*` ones). I don't see what having it in C instead of C++ changes.

_If this is really needed_, there **must** be a PR for this in the next few days (maybe until Wednesday). Otherwise there won't be time to review this.

> kdb should use gopts

Again why?

Here, I see even less reason, why this must be done before the merge. Rewriting `kdb` to use `gopts` could have been done without `new-backend` entirely.

> As discussed with atmaxinger there should be tests for all the hooks.

Sure, but again why before the merge?

For `spec`, `gopts` and notifications at least there are tests that indirectly require them to work. If those tests succeed it should be good enough. We can always add tests explicitly for the hooks in a few weeks when `new-backend` is merged.

**Comment by markus2330** [comment]
> Why is this needed for the merge?
> Again why?

Initially you said that the tooling in the new-backend branch is broken. Nevertheless, this work is at least indirectly needed anyway. If nobody works with new-backend, it is untested/unreviewed. To base some work on it, gives much higher chances that we merge something that works for more than one person. As now several people worked on the new-backend branch and no unknown fatal error was reported so far I came to the conclusion that new-backend is actually ready to be merged very soon after all, hence this issue.

But yes, I agree. Above points (with exception of #4487 and some documentation work) are actually not blockers for new-backend to be merged. If these PRs do not get ready in time, we can merge new-backend also without these changes. Of course this would be a pity and double-work for these authors, thus I wrote this issue for further commitment to be on-time.

In general I hope that as many big changes as possible come in before FLOSS starts, not all necessarily via new-backend (e.g. also #4426).

@kodebach To give back the question: what do you think are must-changes before merging? Now better knowing what my intention is, do you think something is missing in the list above?

**Comment by kodebach** [comment]
> you said that the tooling in the new-backend branch is broken. 

That must have been a long time ago. 841b096f4fdac52160b0277df9f6289d20678c58 is almost a year old. Since around then `kdb mount` should work well enough for the most cases (assuming you want `backend` as your backend plugin)

> Of course this would be a pity and double-work for these authors, thus I wrote this issue for further commitment to be on-time.

Yes, of course as much work as possible should be done before the merge. But my original goal for #4483 (which moved to this issue) was to have _hard_ requirements to decide when merging is possible.

> e.g. also https://github.com/ElektraInitiative/libelektra/pull/4426

Since I'm supposed to update docs for `new-backend`, it's unlikely that I'll finish #4426. I assume you don't want to merge such a big change during FLOSS, so I think it's better to leave the new API stuff until after FLOSS.

>  what do you think are must-changes before merging?

- Docs can certainly be updated after the merge.
  I already started in #4484 and I'll try to update more within the next weeks, but I wouldn't wait with the merge just because of the docs.
- Fixing tests before the merge would of course be preferable. But not everything must be done. Especially bindings and some less important plugins could stay broken in the beginning. Also updating the disabled tests and adding tests for untested stuff would be nice, but I don't think it's a must have for the merge as long as it is done quickly afterwards.
- The rebase/merge with changes from `master` _must_ be done.
- All the other changes (new kdb, proper mount tools, hooks, extra tests) are nice to have, but if there is no PR in the next few days (to give time for reviews) or the PRs aren't merged into `new-backend` before the next release, I would just merge `new-backend` without those changes. Better to have the current stuff that seems to work, than to have something new that might be broken.

**Comment by markus2330** [comment]
> That must have been a long time ago. https://github.com/ElektraInitiative/libelektra/commit/841b096f4fdac52160b0277df9f6289d20678c58 is almost a year old.

It was around 2022-05-11 (new backend meeting), so probably there was some misunderstanding.

> was to have hard requirements to decide when merging is possible.

I think the only left-over hard requirements are that all tests must pass and that there are no merge-conflicts.

> Docs can certainly be updated after the merge.

On the one hand it is impossible to make it a hard requirement in this situation (nobody can do a careful review of the whole docu but nearly everything is effected) but on the other hand PRs never should make docu wrong, the same as they never should break CI/tests.

> but if there is no PR in the next few days (to give time for reviews)

I also would like to have the PRs much earlier.

> Better to have the current stuff that seems to work, than to have something new that might be broken.

Yes.

**Comment by kodebach** [comment]
> all tests must pass

Like I said, I think we can loosen this requirement a bit. Tests for bindings and rarely used plugins are not so important.

Also there are few questions left:
- Are commented out or disabled tests acceptable?
- Must the tests for old data structures, etc. be replaced by new ones?

IMO disabled tests are fine and we don't have to have all the new tests before the merge. I say this because AFAIK all the disabled/missing tests are for low-level stuff that is indirectly tested through other tests. So we know it generally works, but there might be edge cases that would only be caught in specific tests.

Here's a list of what I think about the tests failing in https://cirrus-ci.com/task/6315350697443328:

- `testtool_backend`, `test_tools_plugindatabase.py`
   Would be nice to have fixed. But all the code this tests will soon be replaced anyway. Also `kdb mount` clearly works, because lot's of `testshell_markdown_*` tests work.
- `testlib_notification`
   Can be ignored, notifications aren't fully implemented.
- `test_getenv`
   Not sure whats wrong, but IMO the "intercept" stuff is not so important.
- `testmod_resolver`
   IIRC the test fails because of this commit 4c8727173bdab07e2f9c2284d7e1c8727d95287c, which moved the initalization of the internal data of `resolver` from `open` to `get`. However, the tests expect the initalization to happen in `open`. However, since the resolver is the basis of so many things and so many other tests work, I don't think this must be fixed before the merge.
- `testscr_check_doc`, `testscr_check_plugins`
   I'll fix those in #4484.
- `testjna_gradle`, `test_kdb.py`, `test_gopts.py`, `test_kdb.lua`
   Binding, not so important IMO
- `testshell_markdown_blacklist`, `testshell_markdown_conditionals`, `testshell_markdown_email`, `testshell_markdown_ipaddr`, `testshell_markdown_length`, `testshell_markdown_range`, `testscr_check_gen`, `testshell_markdown_tutorial_validation`
   All these fail because `spec` doesn't work correctly since #4471 AFAICT. This should be fixed, by making `spec` work properly.
- `testshell_markdown_blockresolver`, `testshell_markdown_multifile`, `testshell_markdown_shell`, `testshell_markdown_tutorial_crypto`
  Less used plugin, not so important IMO. But all of them seem to be acting as a resolver and therefore need to implement the new commit function.
- `testshell_markdown_cpptemplate`
  Not sure what's going on there, but I don't think fixing the template test is super important.
- `testscr_check_kdb_internal_suite`, `testscr_check_resolver`
  Not sure, if it's easy fix it, otherwise ignore it, doesn't seem super important considering most other things work.
- `testshell_markdown_cache_shelltests`
   Not sure what's going on here, but if it is related to the cache (as the name suggests): Ignore it, cache is not implemented.
- `testkdb_error`, `testkdb_nested`, `testkdb_simple`
   These should definitely work. And since they worked before #4471 it should be possible to fix them.


So actually IMO the only hard requirements regarding tests is: Fix the things that worked before #4471.

**Comment by markus2330** [comment]
> Are commented out or disabled tests acceptable?

Yes, absolutely. The task FLOSS T2 is about reenabling tests. It is only important that these tests are clearly marked as DISABLED and also listed in doc/todo/TESTING. Ideally there should be some hint what is to do etc. more or less what you wrote above in https://github.com/ElektraInitiative/libelektra/issues/4485#issuecomment-1259494517
Please put this information in doc/todo/TESTING ordered by priority.

>  Must the tests for old data structures, etc. be replaced by new ones?

Yes, but this is #4483, so it is not a blocker to get this merged.

Continuing from https://github.com/ElektraInitiative/libelektra/pull/4484#issuecomment-1273107180

@kodebach said:
> IMO ASAP is basically now

Yes. What is your timeline? Can you do more work today/tomorrow so that we can have it at least mostly reviewed and tested?

> It just needs to be updated. I didn't have time for that, and I believe that updating it could also be done by someone else who has read the newly added docs.

I totally disagree. The documentation simply doesn't give the needed high-level viewpoint at the moment, which makes it very difficult for newcomers to understand what is going on. Even I struggle with understanding the PR. My only advantage is that you told me quite a few things (e.g. KeySet instead of Trie) so I have some ideas what is going on. But for a newcomer it must be terrible. IMHO it is understandable that they are lost and cannot contribute much. This should also change ASAP (but a bit later ASAP than merging).

> Of course, but if everyone does a bit, it's much easier for everyone.

I absolutely agree! There might be other reasons too, but I am afraid that the missing understandable introductions to the new-backend is the main reason.

@flo91 @atmaxinger @Eiskasten and @hannes99 can you give us some insight what the problem is? Can you do more contributions for new-backend soon?

**Comment by atmaxinger** [comment]
Well for me at least, yes I have time to do contributions for new-backend. But the big problem is: my knowledge of this branch consists entirely of the documentation that I read and the countless things @kodebach told me when I ran into some troubles.

The few things I have contributed to the branch and have a good understanding about, i.e. the hooks system, has already been documented by me.

That said, I'd gladly help, but I just don't know where to even begin.

**Comment by kodebach** [comment]
> What is your timeline? Can you do more work today/tomorrow so that we can have it at least mostly reviewed and tested?

I can definitely finish fixing/disabling the tomorrow (Tuesday). How of your review I can incorporate before the meeting on Wednesday I'm not sure. I think we should discuss the timeline on Wednesday, but definitely merge #4187 this week. I don't think there is anything really big and blocking left.

> Even I struggle with understanding the PR.

This may also be because you have a lot of knowledge, about how things worked before. In the end, other than the things that are documented everything should behave as before. The internal code is just very different.

> I just don't know where to even begin.

It is kind of the same for me. I know a few corners where no up-to-date docs exist (e.g. `src/plugins/backend/README.md`), but generally I think I have documented most things. The only thing I can do is write about the same information on different levels. But (and I've said this before) I'm not sure the results would be very good. _Because_ I have so much internal knowledge I have to be very careful to not overexplain things. It would also be a good test for the existing docs, if someone read them and tried to write more high-level docs based on those. Because if they misunderstand something or do not understand something at all, then the existing docs need to be improved as well.

> KeySet instead of Trie

This for example, is something that nobody should need to know to understand Elektra. It helps _you_ to know, because you knew the previous implementation, but without prior knowledge it should really not matter how `struct _KDB` is organised unless you want to change it. 

> the missing understandable introductions to the new-backend

What exactly do you want there? What are "introductions to the new-backend"? new-backend is just a branch, once it's merged it's just Elektra. I think we have quite a few introductions to Elektra already.

From a user perspective, everything should be basically the same right now. There is only a single usable backend plugin and `kdb mount` hasn't changed. The only real difference I can think of, is that the "too many plugins" error is gone.

From a developer perspective, the application APIs haven't really changed. The few minor changes that exist are already documented (e.g. in doxygen `kdbGet`/`kdbSet`). For developing plugins everything is also mostly the same still. Resolver plugins are very slightly different (they get the relative path in a different way). Backend plugins are completely new, but it's also document what those should do.

From a contributor perspective, I don't really know what else I could say. There is lots of low-level docs of what `libelektra-kdb` does.

Can you give me a concrete example of something that's missing? A concrete topic I should write about?

**Comment by markus2330** [comment]
Can we close this or is everything left to be done before the release?

**Comment by kodebach** [comment]
I think everything is done

**Comment by markus2330** [comment]
Is there some FIXME/TODO that shows a serious bug? (Users can trigger it and would experience crashes, data loss, ..?) If yes, please open issues for them.

**Comment by kodebach** [comment]
No, AFAIK it's only:
- missing/outdated documentation
- disabled tests
- missing, but optional features like `cache`
- some minor things where the implementation doesn't yet match what's written in `doc/dev/kdb-operations.md` et al., e.g. I'm not sure `poststorage` plugins are executed for `default:/` keys during `kdbGet`.

However, there is one thing that might still prevent a release: AFAIK we don't have a migration path from the old `system:/elektra/mountpoints` layout to the new one yet.
